File: scala-java/src/main/java/at/twinformatics/scalademo/java/Generics.java
package at.twinformatics.scalademo.java;

/**
 * Example for defining and using generic types in Java. Java allows only
 * use-side variance declaration.
 */
public class Generics {

    interface Iterator<T> {
        boolean hasNext();
        T next();
    }

    Iterator<String> strings = null;
    Iterator<?> objects = strings;
    Iterator<Integer> integers = null;

    void iterate(Iterator<? extends Number> it) {}
    <T extends Number> void iterate2(Iterator<T> it) {}

    { iterate(integers); }
    { iterate2(integers); }

}

File: scala-java/src/main/scala/at/twinformatics/scalademo/scala/Generics.scala
package at.twinformatics.scalademo.scala

/**
  * Example on how to use declaration side and use side generic variance
  * definition.
  */
object Generics {

    trait Iterator[+T] {
        def hasNext: Boolean
        def next: T
    }

    val strings: Iterator[String] = null
    val objects: Iterator[AnyRef] = strings
    val integers: Iterator[Integer] = null

    def iterate(it: Iterator[Number]): Unit = ()
    def iterate2[T <: Number](it: Iterator[T]): Unit = ()

    iterate(integers)
    iterate2(integers)
}
